### ISSUE TITLE
css wrong path with rollup-plugin-svelte-hot 0.11

### ISSUE THREAD

**Issue by frederikhors** [rixo/svelte-template-hot#18]
Using this in rollup.config.js:

```js
css: css => {
  css.write('public/build/bundle.css')
}
```

is not ok with `rollup-plugin-svelte-hot: 0.11.0`

because it creates a fake `public/build/bundle.css` using `npm run dev` and a wrong `public/build/public/build/bundle.css` using `npm run build`.

Using this code instead:

```js
css: css => {
  css.write('bundle.css')
}
```

creates a fake `bundle.css` in root (which is wrong) with `npm run dev` and a good `public/build/bundle.css` with `npm run build`.

**Comment by rixo** [comment]
Interesting. I'm pretty sure it was working before. But I encountered exactly the same thing in svench-component-template [just yesterday](https://github.com/rixo/svench-component-template/commit/5c90ea36983bc0ad61a663b9f4d54b33c36d3a14).

Have we recently merged a breaking change from upstream without realizing it? Or did we introduce the breaking change ourselves when resolving conflicts?

I think it'd be worth tracking where this difference in behaviour originate by searching in official rollup-plugin-svelte's changelog, PRs and all... to get a complete understanding of the situation.

Right now, I guess we just need to be pragmatic and change this path in master. This is a template, not a lib, so we don't really have a notion of breaking change here. People are expected to clone it once and move on their own from this point, the template is not a cleanly versioned dep of their project...

**Comment by frederikhors** [comment]
The breaking change in `rollup-plugin-svelte` is here: https://github.com/sveltejs/rollup-plugin-svelte/blob/master/CHANGELOG.md#600.

I'm trying to understand what to do in our `rollup-plugin-svelte-hot`.

**Comment by rixo** [comment]
Oh yes indeed, it was a major. Well `rollup-plugin-svelte-hot` released this as a 0.x bump, which means breaking, so we're good there.

**Comment by rixo** [comment]
OK we're good on the breaking version, but the bug was there. Should be fixed with `rollup-plugin-svelte-hot@0.11.1`. 

I've also updated this template with the new way. Unfortunately Rollup and Nollup seem to disagree on how to resolve `emitAsset`, at least with `output.file` (not tested `output.dir`), so I've put a condition with specific handling of Nollup for now.

Can you confirm on your side the behaviour is as you expected with the last version of rollup-plugin-svelte-hot?

**Comment by frederikhors** [comment]
Oh wow! We were working on the same thing! I just created a PR! LOLOL!!!!!!

**Comment by frederikhors** [comment]
Let me see your changes.

**Comment by frederikhors** [comment]
Surely yours is better! I'm gonna close it.

**Comment by rixo** [comment]
Yeah the bug was mine, not upstream. In hot mode, the plugins write a blank file where the official one would have written the CSS to avoid a 404, so this has to be adapted to follow the new behavior.

**Comment by frederikhors** [comment]
I tried. It works. Thanks!

**Comment by frederikhors** [comment]
I have to reopen this. It's not working!

**Comment by rixo** [comment]
What command is not working?

**Comment by frederikhors** [comment]
The build command works.

The dev doesn't create file in my project. I'm investigating.

**Comment by frederikhors** [comment]
Wow! It's crazy it's the same with this template too, but the browser finds the file when in dev.

**Comment by frederikhors** [comment]
As you can see in dev che file is not written.

In this template the css is called with this: 

`<link rel='stylesheet' href='/build/bundle.css'>`

In my project I'm calling it with:

`<link rel='stylesheet' href='/bundle.css'>`

but it's the same: it doesn't create the file in dev.

**Comment by rixo** [comment]
Nollup (used with default `dev` command) "writes" files in memory so you don't see them on the disk, but it still serves them to the browser.

Unfortunately, it differs slightly from Rollup in how it resolves the emitted file path with an `output.file`. That's why I've had to add this condition in the Rollup config:

~~~js
        css.write(isNollup ? 'build/bundle.css' : 'bundle.css')
~~~

Not sure how this applies to your own project ^^

**Comment by frederikhors** [comment]
Ok I changed this template to reproduce it: https://github.com/frederikhors/svelte-template-hot/tree/issue-18

Changes: https://github.com/frederikhors/svelte-template-hot/commit/209058befe3c9c210bae5cbb5c18714803d12699

**Comment by frederikhors** [comment]
With `rollup-plugin-svelte-hot v10` the css file is created not only in memory! On disk! And my browser finds it.

**Comment by rixo** [comment]
Yes, the breaking change was to use Rollup's `emitFile` instead of direct `fs.writeFile`. I've mirrored this change in last version. With fs, the file was written to fs in any case. Now, Nollup doesn't write to disk and it processes the `emitFile` itself, so the file is not generated at all anymore with Nollup, this is expected.

Nollup is still expected to have the file in memory and to serve it to browser. The problem is there's a mismatch in how Nollup computes the complete path of the file vs Rollup. Hence the special condition I've added in the config for now (should probably be harmonized at Nollup level eventually).

Unfortunately, to debug this I need to put a `console.log` in Nollup's code, so it's not very straightforward... It's the request handler at the end of the dev-middleware.js file. There's a `files` object in there, of which the keys give you the files paths as Nollup sees them.

**Comment by rixo** [comment]
Any progress on this? Need a hand? Can close?

**Comment by frederikhors** [comment]
Oh. Please don't close it. Still a problem. 🥺

**Comment by frederikhors** [comment]
My problem was this line in my `rollup.config.js`:

```js
copy({ targets: [{ src: ['files/*'], dest: 'final' }], copyOnce: true })
```

[which is using by default](https://github.com/vladshcherbin/rollup-plugin-copy#hook) the `buildEnd` Rollup's hook.

I'm using the `buildStart` now and it works good.

Thanks a lot! :smile:

**Comment by frederikhors** [comment]
I spoke too soon.

The problem came back as soon as I started working again and saved a .svelte file.

I am very sad now.

😭

**Comment by rixo** [comment]
Oh, don't be sad ;) So, you're using `output.dir` instead of `output.file`. Actually, I believe this is better supported by Nollup, and is not impacted by the problem I have worked around in the template, with the latest versions of `rollup-plugin-svelte-hot`. For some reasons, it's somewhat harder for Nollup to decipher the URL paths with `output.file` than `output.dir`...

I've tried your `issue-18` branch and did some debugging in Nollup. Apparently, with `output.dir` there's no discrepancy with Rollup (which is what is expected!), so you don't need any special handling. With just the following, I seems to have both `yarn:dev` (Nollup), and `yarn build` working:

```js
      css: css => {
        css.write('bundle.css')
      },
```

However, `yarn dev:rollup` is not working because you've changed the output / public directory... This can be fixed by disabling `sirv` and relying on `rollup-plugin-hot` internal dev server to serve everything.

Disable `sirv` in hot mode:

```js
    isDev && !isHot && serve(),
```

Have `rollup-plugin-hot` serves from dirs `public` and `final` on port 5000:

```js
    hmr({
      public: 'final',
      port: 5000,
      mount: {
        public: '/',
        final: '/',
      },
      // ...
    }),
```

Does this work for you?

**Thoughts**

- Code splitting / `output.dir` should be illustrated by this template. This is like the number 1 query in the feedback I've got. Don't know what form it should take though... A dedicated template? A branch? What would be practical?

- Very tempted to drop `sirv` altogether in dev mode (i.e. watch). It is completely oriented toward prod setup and performance, and is causing problems in dev, both here and the official template (try saving your rollup.config.js while Rollup is running to spawn sirv zombie processes -- first one of them squatting the port 5000 where you expect to see your most current result)... Here we have an OK dev server in rollup-plugin-hot, I think we should leverage it more and avoid those issues. Any opinion on this?

**Comment by frederikhors** [comment]
Thanks for your answer.

As you noticed it works with:

```js
css: css => {
    css.write("bundle.css", false)
}
```

But as soon as you make a single change and refresh the page (total refresh, not hot-refresh) we have the `bundle.css` error again in console:

```
Refused to apply style from 'http://localhost:5000/bundle.css' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
```

### For questions

1. I think we can just write a comment for now. If we fix this error there are no others and I am happily using it.

2. I would eliminate sirv.

![image](https://user-images.githubusercontent.com/41120635/98472235-169b2280-21f2-11eb-8803-0715268f2180.png)

**Comment by rixo** [comment]
Oh indeed. Traced this back into Nollup. It clears its files on change, and refill them with updated files (I guess).

`rollup-plugin-svelte-hot` calls the `svelte.css` function once when started, specifically to avoid this 404.

So, the fix belongs into `rollup-plugin-svelte-hot`, in the category "enhanced Nollup support".

In any event, it's a cosmetic error, it doesn't have a real impact on functioning, so you can safely ignore it for now.

**Comment by frederikhors** [comment]
Yes, it's just boring to see in the browser console.

However I would leave this issue open until we fix it.

**Comment by rixo** [comment]
This should be fixed in `rollup-plugin-svelte-hot@0.11.2`. Can you confirm?

**Comment by frederikhors** [comment]
It works! 😄
